File: SearchingAndSorting/questions.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

int binarySearch(vector<int> &arr, int data)
{
    int n = arr.size(), si = 0, ei = n - 1;

    while (si <= ei) // equal hone tk chlaate hai as tb mid check nhi hua hota
    {
        int mid = (si + ei) / 2; // use si + (ei - si)/2 when chances of overflowing
        if (arr[mid] == data)
            return mid;
        else if (arr[mid] < data)
            si = mid + 1;
        else
            ei = mid - 1;
    }

    return -1;
}

int firstIndex(vector<int> &arr, int data)
{
    int n = arr.size(), si = 0, ei = n - 1;
    while (si <= ei)
    {
        int mid = (si + ei) / 2;
        if (arr[mid] == data)
        {
            if (mid - 1 >= 0 && arr[mid - 1] == data)
                ei = mid - 1;
            else
                return mid;
        }
        else if (arr[mid] < data)
            si = mid + 1;
        else
            ei = mid - 1;
    }
    return -1;
}

int lastIndex(vector<int> &arr, int data)
{
    int n = arr.size(), si = 0, ei = n - 1;
    while (si <= ei)
    {
        int mid = (si + ei) / 2;
        if (arr[mid] == data)
        {
            if (mid + 1 < n && arr[mid + 1] == data)
                si = mid + 1;
            else
                return mid;
        }
        else if (arr[mid] < data)
            si = mid + 1;
        else
            ei = mid - 1;
    }
    return -1;
}

//Leetcode 34 -> Find First and Last Position of Element in Sorted Array
vector<int> &searchRange(vector<int> &arr, int data)
{
    return new vector<int> &{firstIndex(arr, data), lastIndex(arr, data)};
}

//Important
int insertLocation(vector<int> &arr, int data)
{
    int n = arr.size(), si = 0, ei = n - 1;
    while (si <= ei)
    {
        int mid = (si + ei) / 2;

        if (arr[mid] <= data)
            si = mid + 1;
        else
            ei = mid - 1;
    }

    return si;
}

int perfectPosOfElement(vector<int> &arr, int data)
{
    int insertPos = insertLocation(arr, data);
    int lastIndex = insertPos - 1;

    return (lastIndex >= 0 && arr[lastIndex] == data) ? lastIndex : insertPos;
}

//firstIndex using the concept used in insertLocation function
int firstIndex_02(vector<int> &arr, int data)
{
    int n = arr.size(), si = 0, ei = n - 1;
    while (si <= ei)
    {
        int mid = (si + ei) / 2;
        if (arr[data] < mid) // equal to mei ei ko move krenge
            si = mid + 1;
        else
            ei = mid - 1;
    }

    return si < n && arr[si] == data ? si : -1;
}

//LastIndex using the concept used in insertLocation function
int lastIndex_02(vector<int> &arr, int data)
{
    int n = arr.size(), si = 0, ei = n - 1;
    while (si <= ei)
    {
        int mid = (si + ei) / 2;
        if (arr[mid] <= data) // equal to pe bhi si ko move krenge
            si = mid + 1;
        else
            ei = mid - 1;
    }

    si--;
    return si >= 0 && arr[si] == data ? si : -1;
}

//Return the nearest element present in arr to data
// Always aftr the while loop -> arr[ei] <= data < arr[si]
int nearestElement(vector<int> &arr, int data)
{
    int n = arr.size(), si = 0, ei = n - 1;

    if (n == 0)
        return 0;

    if (data <= arr[0] || data >= arr[n - 1])
        return data <= arr[0] ? arr[0] : arr[1];

    while (si <= ei)
    {
        int mid = (si + ei) / 2;
        if (arr[mid] <= data)
            si = mid + 1; // si is at ceil value after while loop
        else
            ei = mid - 1; // ei is at floor value after while loop
    }

    return data - arr[ei] <= arr[si] - data ? arr[ei] : arr[si];
}

//Leetcode 74 -> Search a 2D matrix
bool searchMatrix(vector<vector<int> > &matrix, int target)
{
    int n = matrix.size(), m = matrix[0].size(), si = 0, ei = n * m - 1;

    while (si <= ei)
    {
        int mid = (si + ei) / 2;
        int r = mid / m, c = mid % m;

        if (matrix[r][c] == target)
            return true;
        else if (matrix[r][c] < target)
            si = mid + 1;
        else
            ei = mid - 1;
    }
    return false;
}

//Leetcode 240 -> Search a 2D Matrix II
bool searchMatrix(vector<vector<int> > &matrix, int target)
{
    int n = matrix.size(), m = matrix[0].size(), r = n - 1, c = 0;

    while (r >= 0 && c < m)
    {
        int ele = matrix[r][c];
        if (ele == target)
            return true;

        else if (ele < target)
            c++;
        else
            r--;
    }
    return false;
}

//Leetcode 167 -> Two Sum II
vector<int> twoSum(vector<int> &arr, int target)
{
    int si = 0, ei = arr.size() - 1;

    while (si < ei)
    {
        int sum = arr[si] + arr[ei];
        if (sum == target)
            return {si + 1, ei + 1};
        if (sum < target)
            si++;
        else
            ei--;
    }
    return {};
}

//Now return all the possible unique pairs that add upto target
vector<vector<int> > twoSum_AllPairs(vector<int> &arr, int target, int si, int ei)
{
    vector<vector<int> > ans;
    while (si < ei)
    {
        int sum = arr[si] + arr[ei];
        if (sum == target)
        {
            ans.push_back({arr[si], arr[ei]});

            si++;
            ei--;
            while (si < ei && arr[si] == arr[si - 1]) // si and ei ko unique ele pe pohocha dia
                si++;
            while (si < ei && arr[ei] == arr[ei + 1])
                ei--;
        }
        else if (sum < target)
            si++;
        else
            ei--;
    }
    return ans;
}

//Leetcode 15 -> 3Sum -> Uses TwoSumAllPairs
void prepareAns(vector<vector<int> > &smallAns, vector<vector<int> > &ans, int fixedEle)
{
    for (vector<int> arr : smallAns)
    {
        vector<int> ar;
        ar.push_back(fixedEle);
        for (int ele : arr)
            ar.push_back(ele);
        ans.push_back(ar);
    }
}

vector<vector<int> > threeSum(vector<int> &arr, int target, int si, int ei)
{
    vector<vector<int> > ans;
    for (int i = si; i < ei;)
    {
        vector<vector<int> > smallAns = twoSum_AllPairs(arr, target - arr[i], i + 1, ei);
        prepareAns(smallAns, ans, arr[i]);
        i++;
        while (i < ei && arr[i] == arr[i - 1])
            i++;
    }
    return ans;
}

//Leetcode function
vector<vector<int> > threeSum(vector<int> &arr)
{
    int n = arr.size();
    sort(arr.begin(), arr.end());
    return threeSum(arr, 0, 0, n - 1);
}

//Leetcode 18 -> 4Sum
//Uses 3Sum (which in turn uses 2Sum) and prepareAns
vector<vector<int> > fourSum(vector<int> &arr, int target, int si, int ei)
{
    vector<vector<int> > ans;
    for (int i = si; i < ei;)
    {
        vector<vector<int> > smallAns = threeSum(arr, target - arr[i], i + 1, ei);
        prepareAns(smallAns, ans, arr[i]);
        i++;
        while (i < ei && arr[i] == arr[i - 1])
            i++;
    }
    return ans;
}

//Leetcode function
vector<vector<int>> fourSum(vector<int> &arr, int target)
{
    int n = arr.size();
    sort(arr.begin(), arr.end());
    return fourSum(arr, target, 0, n - 1);
}

//Generic function for kSum
vector<vector<int>> kSum(vector<int> &arr, int target, int k, int si, int ei)
{
    if (k == 2)
        return twoSum_AllPairs(arr, target, si, ei);

    vector<vector<int>> ans;
    for (int i = si; i < ei;)
    {
        vector<vector<int> > smallAns = kSum(arr, target - arr[i], k - 1, i + 1, ei);
        prepareAns(smallAns, ans, arr[i]);
        i++;
        while (i < ei && arr[i] == arr[i - 1])
            i++;
    }

    return ans;
}

//Return all the pairs that add upto target one ele should be from nums1 and the other from nums2
int twoSumCount(vector<int>& nums1,vector<int>& nums2,int target)
{
    unordered_map<int,int> map;
    for(int ele : nums1)
        map[ele]++;  // freq of ele in nums1 array

    int count = 0;
    for(int ele : nums2)
    {
        if(map.find(target - ele) != map.end())
            count += map[target - ele];
    }

    return count;
}

//Leetcode 454 -> FourSum II -> Done similarly as twoSumCount
int fourSumCount(vector<int>& nums1, vector<int>& nums2, vector<int>& nums3, vector<int>& nums4) 
{
    unordered_map<int,int> map;

    for(int e1 : nums1)
        for(int e2 : nums2)
            map[e1+e2]++;

    int count = 0, target = 0;
    for(int e3 : nums3)
        for(int e4 : nums4)
            if(map.find(target - (e3 + e4)) != map.end())
            count += map[target - (e3 + e4)];

    return count;
}

int fourSumCount_02(vector<int>& nums1, vector<int>& nums2, vector<int>& nums3, vector<int>& nums4)
{
    vector<int> A;
    vector<int> B;
    for(int e1 : nums1)
        for(int e2 : nums2)
            A.push_back(e1 + e2);
    
    for(int e3 : nums3)
        for(int e4 : nums4)
            B.push_back(e3 + e4);

    return twoSumCount(A,B,0);
}

//Leetcode 658 -> Find K Closest Elements
int insertPosition(vector<int>& arr,int data)
{
    int n = arr.size(), si = 0, ei = n-1;
    while(si <= ei)
    {
        int mid = (si + ei) / 2;
        if(arr[mid] <= data)
            si = mid+1;
        else
            ei = mid-1;
    }
    return si;  // last mei si insertPosition mei pohoch jaata hai
}

//Leetcode function
vector<int> findClosestElements(vector<int>& arr, int k, int x) 
{        
    int n = arr.size();
    int idx = insertPosition(arr,x);
        
    int lr = max(0,idx - k);
    int rr = min(n-1,idx + k);
        
    while((rr - lr) + 1 > k)
    {
        if(x - arr[lr] > arr[rr] - x)
            lr++;
        else 
            rr--;
    }
    return {arr.begin() + lr,arr.begin() + rr + 1};
}

File: SearchingAndSorting/questions.java
import java.util.Arrays;
import java.util.ArrayList;

public class questions{

public static int binarySearch(int[] arr,int data)
{
    int n = arr.length, si = 0, ei = n - 1;

    while(si <= ei)
    {
        int mid = (si + ei) / 2;
        if(arr[mid] == data)
            return data;

        else if(arr[mid] < data)
            si = mid+1;
        else
            ei = mid-1;
    }

    return -1;
}

public int firstIndex(int[] arr,int data)
{
    int n = arr.length, si = 0, ei = n-1;
    while(si <= ei)
    {
        int mid = (si + ei) / 2;
        if(arr[mid] == data)
        {
            if(mid - 1 >= 0 && arr[mid-1] == data)
                ei = mid-1;
            else
                return mid;
        }
        else if(arr[mid] < data)
            si = mid + 1;
        else
            ei = mid - 1;
    }
    return -1;
}

public int lastIndex(int[] arr,int data)
{
    int n = arr.length, si = 0, ei = n - 1;
    while(si <= ei)
    {
        int mid = (si + ei) / 2;
        if(arr[mid] == data)
        {
            if(mid + 1 < n && arr[mid+1] == data)
                si = mid+1;
            else
                return mid;
        }
        else if(arr[mid] < data)
            si = mid+1;
        else
            ei = mid-1;
    }
    return -1;
}

//Leetcode 34 -> Find First and Last Position of Element in Sorted Array
public int[] searchRange(int[] arr, int data) 
{
    return new int[]{firstIndex(arr,data),lastIndex(arr,data)};
}

//Leetcode 35 -> Search Insert Position
public int searchInsert(int[] arr, int data) 
{
    int n = arr.length, si = 0, ei = n-1;
    while(si <= ei)
    {
        int mid = (si + ei) / 2;
            
        if(arr[mid] <= data)
            si = mid + 1;
        else
            ei = mid-1;
    }
        
    int lastIndex = si - 1;
    int insertPos = si;
    return (lastIndex >= 0 && arr[lastIndex] == data) ? lastIndex : insertPos;
}

//Inversion Count -> (GFG)
// Returns inversion count across two Arrays and sort the array at that range(merge sort)
public static long inversionAcrossArray(long[] arr,int l,int r,int mid,long[] sortedArray)
{
    int lsi = l, lei = mid;
    int rsi = mid+1, rei = r;

    int k = 0;
    long count = 0;
    while(lsi <= lei && rsi <= rei)
    {
        if(arr[lsi] > arr[rsi])
        {
            count += (lei - lsi + 1);
            sortedArray[k++] = arr[rsi++];
        }
        else
            sortedArray[k++] = arr[lsi++];
    }

    while(lsi <= lei) // agr dono mei se koi arr bach gya ho end mei to as it is copy
        sortedArray[k++] = arr[lsi++];
    while(rsi <= rei)
        sortedArray[k++] = arr[rsi++];   

    k = 0;
    for(int i = l; i <= r; i++)
        arr[i] = sortedArray[k++];  // sortedArr copied into original arr to make changes in the real arr.
    
    return count;
}

// Main recursive function
public static long inversionCount(long[] arr,int si,int ei,long[] sortedArray)
{
    if(si >= ei)
    return 0;

    int mid = (si + ei) / 2;

    long leftInversionCount = inversionCount(arr,si,mid,sortedArray);  // Inversion count within left subArray
    long rightInversionCount = inversionCount(arr,mid+1,ei,sortedArray); 

    return (leftInversionCount + rightInversionCount + inversionAcrossArray(arr,si,ei,mid,sortedArray));
}

//GFG function
public static long inversionCount(long arr[], long N)
{
    if(N <= 1)
    return 0;

    long[] sortedArray = new long[(int)N];
    return inversionCount(arr,0,(int)N-1,sortedArray);
}

//Leetcode 33 -> Search in Rotated Sorted Array
public int search(int[] arr, int target) 
{
    int si = 0, ei = arr.length-1;
        
    while(si <= ei)
    {
        int mid = (si + ei) / 2;

        if(arr[mid] == target)
            return mid;
        else if(arr[si] <= arr[mid])
        {
            if(arr[si] <= target && target < arr[mid])
                ei = mid-1;
            else
                si = mid+1;
        }
        else
        {
            if(arr[mid] < target && target <= arr[ei])
                si = mid+1;
            else
                ei = mid-1;
        }
    }
    return -1;
}

//Leetcode 81 -> Search in Rotated Sorted Array -> Now duplicate values are also present in the arr
public boolean search(int[] arr, int target) 
{
    int si = 0, ei = arr.length-1;
        
    while(si <= ei)
    {
        int mid = (si + ei) / 2;
        
        if(arr[mid] == target || arr[si] == target)
            return true;
        else if(arr[si] < arr[mid])
        {
            if(arr[si] <= target && target < arr[mid])
                ei = mid-1;
            else
                si = mid+1;
        }
        else if(arr[mid] < arr[ei])
        {
            if(arr[mid] < target && target <= arr[ei])
                si = mid+1;
            else
                ei = mid-1;
        }
        else
            si++;  // worst case -> O(N) when all ele are duplicate
    }
    return false;
}

//Leetcode 153 -> Find Minimum in Rotated Sorted Array -> Find the pivot ele
public int findMin(int[] arr) 
{
    int n = arr.length, si = 0, ei = n - 1;
    if(arr[si] <= arr[ei])
        return arr[si];
        
    while(si < ei)
    {
        int mid = (si + ei) / 2;
            
        if(arr[mid] < arr[ei])
            ei = mid;
        else if(arr[si] <= arr[mid])
            si = mid + 1;
    }
    return arr[si];
}

//Leetcode 154 -> Find Minimum in Sorted Array II -> Now duplicate values are allowed
public int findMin(int[] arr) 
{
    int si = 0, ei = arr.length-1;
        
    if(arr[si] < arr[ei])
        return arr[si];
        
    while(si < ei)
    {
        int mid = (si + ei) / 2;
            
        if(arr[mid] < arr[ei])
            ei = mid;
        else if(arr[mid] > arr[ei])
            si = mid+1;
        else
            ei--;
    }
    return arr[si];
}

//Leetcode 167 -> Two Sum II - Input Array is Sorted
public int[] twoSum(int[] arr, int target) 
{
    int si = 0, ei = arr.length-1;
    
    while(si < ei)
    {
        int sum = arr[si] + arr[ei];
        if(sum == target)
            return new int[]{si+1,ei+1};
        if(sum < target)
            si++;
        else
            ei--;
    }
    return new int[0];
}

//Leetcode 658 -> Find K Closest Elements
public int insertPosition(int[] arr,int data)
{
    int n = arr.length, si = 0, ei = n-1;
    while(si <= ei)
    {
        int mid = (si + ei) / 2;
        if(arr[mid] <= data)
            si = mid+1;
        else
            ei = mid-1;
    }
    return si;  // last mei si insertPosition mei pohoch jaata hai
}

//Leetcode function
public List<Integer> findClosestElements(int[] arr, int k, int x) 
{
    ArrayList<Integer> ans = new ArrayList<>();
    for(int ele : arr)
        ans.add(ele);
        
    int n = arr.length;
    int idx = insertPosition(arr,x);
        
    int lr = Math.max(0,idx - k);
    int rr = Math.min(n-1,idx + k);
    
    if(x <= arr[0])
        return ans.subList(0,k);
    else if(x >= arr[n-1])
        return ans.subList(n-k,n);
    else
    {
        while((rr - lr) + 1 > k)
        {
            if(x - arr[lr] > arr[rr] - x)
                lr++;
            else 
                rr--;
        }
        return ans.subList(lr,rr+1);
    }
}

//Leetcode 300 -> Longest Increasing Subsequence
public int insertPosition(ArrayList<Integer> arr,int data)
{
    int si = 0, ei = arr.size()-1;
    while(si <= ei)
    {
        int mid = (si + ei) / 2;
        if(arr.get(mid) <= data)
            si = mid+1;
        else
            ei = mid-1;
    }

    int insertPos = si;
    int lastIndex = si-1;
        
    return lastIndex >= 0 && arr.get(lastIndex) == data ? lastIndex : insertPos; // if Present then its lastIndex if not then insertPosition
}

//Leetcode function
public int lengthOfLIS(int[] nums)
{
    int n = nums.length;
    if(n <= 1)
        return n;
        
    ArrayList<Integer> list = new ArrayList<>();
    for(int ele : nums)
    {
        int loc = insertPosition(list,ele);
        if(loc == list.size())
            list.add(ele);
        else
            list.set(loc,ele);
    }
    return list.size();
}

//Leetcode 875 -> Koko Eating Bananas
public boolean isPossibleToEat(int[] arr,int hour,int eatingSpeed)
{
    int hr = 0;
    for(int i = arr.length-1; i >= 0; i--)
    {
        hr += Math.ceil(arr[i] / (eatingSpeed * 1.0));
        if(hr > hour)
            return false;
    }
    return true;
}

//Leetcode function
public int minEatingSpeed(int[] piles, int h) 
{
    int si = 1, ei = (int)1e9;
    
    while(si < ei)
    {
        int eatingSpeed = (si + ei) / 2;
        if(!isPossibleToEat(piles,h,eatingSpeed))
            si = eatingSpeed + 1;   // speed kam hai to speed badaani hai
        else
            ei = eatingSpeed;  // nhi to speed kam krke dekhna hai aur chotta ans mil jaaye to
    }
    return si;
}

//Leetcode 1011 -> Capacity to Ship Packages Within D days
public boolean isPossibleToShip(int[] weights,int capacity,int days)
{
    int d = 1, totalWeightPerDay = 0;
    for(int w : weights)
    {
        totalWeightPerDay += w;
        if(totalWeightPerDay > capacity)
        {
            d++;
            totalWeightPerDay = w;   
        }
            
        if(d > days)
            return false;
    }
        
    return true;        
}

//Leetcode function
public int shipWithinDays(int[] weights, int days) 
{
    int maxEle = 0, sum = 0;
    for(int w : weights)
    {
        maxEle = Math.max(maxEle,w);
        sum += w;
    }
            
    int si = maxEle, ei = sum;
        
    while(si < ei)
    {
        int capacity = (si + ei) / 2;
        if(!isPossibleToShip(weights,capacity,days))
            si = capacity + 1;
        else
            ei = capacity;
    }
    return si;
}

//Method 2 -> Now range 1 se 1e7 tk rakhenge 
public boolean isPossibleToShip_02(int[] weights,int capacity,int days)
{
    int d = 1, totalWeightPerDay = 0;
    for(int w : weights)
    {
        if(w > capacity)   // agr present weight capacity se bada hogya to hum usko carry nhi kr skte
            return false;
        totalWeightPerDay += w;
        if(totalWeightPerDay > capacity)
        {
            d++;
            totalWeightPerDay = w;   
        }
            
        if(d > days)
            return false;
    }
        
    return true;        
}

//Leetcode function
public int shipWithinDays_02(int[] weights, int days) 
{
    int si = 1, ei = (int)1e7;  
        
    while(si < ei)
    {
        int capacity = (si + ei) / 2;
        if(!isPossibleToShip_02(weights,capacity,days))
            si = capacity + 1;
        else
            ei = capacity;
    }
    return si;
}

//Max Cake Area Question
public boolean isPossibleToServeCake(int[] radiusArray,int cakeArea,int guests)
{
    int g = 0;
    for(int i = radiusArray.length-1; i >= 0; i--)
    {
        double area = Math.PI * radiusArray[i] * radiusArray[i];
        g += Math.floor(area / cakeArea);
        
        if(g >= guests)
        return true;
    }
    return false;
}

public static double maximumAreaCake(int[] radius,int guests)
{
    int n = radius.length;
    double si = 0.0, ei = 1e7;

    while((ei - si) > 1e-5)
    {
        double cakeArea = (si + ei) / 2.0;

        if(!isPossibleToServeCake(radius,cakeArea,guests))
            ei = cakeArea - 1e-5;
        else
            si = cakeArea;
    }
    return si;
}

//Leetcode 774 -> Minimize Max Distance to Gas Distance(Lintcode 848)
public boolean isItCorrectPenalty(int[] arr,double distance,int k)
{
    int noOfGasStation = 0, n = arr.length;
    for(int i = 1; i < n; i++)
    {
        noOfGasStation += (arr[i] - arr[i-1]) / distance;
        if(noOfGasStation > k)
            return false;
    }
    return true;
}

//Leetcode function
public double minmaxGasDist(int[] stations, int k) 
{
    double si = 0.0, ei = 1e9;
    while((ei - si) > 1e-6)
    {
        double distance = (si + ei) / 2;
        if(!isItCorrectPenalty(stations,distance,k))
            si = distance + 1e-6;
        else
            ei = distance;
    }
    return si;
}